lib/wraft_doc/search/encoder.ex

```diff
@@ -0,0 +1,114 @@
+defprotocol WraftDoc.Search.Encoder do
+  @moduledoc """
+  Protocol defining the interface for converting records to Typesense documents.
+  """
+
+  @doc """
+  Converts a struct to a map format .
+  """
+  @spec to_document(t()) :: map()
+  def to_document(struct)
+end
+
+defimpl WraftDoc.Search.Encoder,
+  for: [
+    WraftDoc.Document.ContentType,
+    WraftDoc.Document.DataTemplate,
+    WraftDoc.Document.Layout,
+    WraftDoc.Document.Theme,
+    WraftDoc.Enterprise.Flow
+  ] do
```

Comment:
Lets not use this pattern since the implementation is different for each type.

While using pattern matching within a `defimpl` to handle multiple types works, it’s not the most idiomatic approach in Elixir. Protocols are designed to abstract behavior cleanly for each type, and grouping multiple types with pattern matching in a single implementation can reduce modularity and make it harder to extend or maintain in the future.

A more idiomatic approach would be to define separate `defimpl` blocks for each type (`ContentType`, `DataTemplate`, etc.). This aligns better with Elixir’s philosophy of clear and explicit behavior for each type and makes the codebase easier to scale and reason about over time